### PR TITLE
new library functions

### PR DESCRIPTION
**Pull request by bradh352**
* ares__buf_split_str: Ability to split a buf object into a C array of C strings
* ares__array_insertdata_{at,last,first}: copy data provided into array member as a helper
* ares_free_array: Free a normal C array with custom callback to free each element

Authored-By: Brad House (@bradh352)

### CODE REVIEW

File: src/lib/str/ares__buf.c [hunk 1]
```diff
@@ -884,6 +892,61 @@ ares_status_t ares__buf_split(ares__buf_t *buf, const unsigned char *delims,
...
+      goto done;
+    }
+  }
+
+done:
+  ares__llist_destroy(list);
+  if (status == ARES_SUCCESS) {
+    *strs = ares__array_finish(arr, nstrs);
+  } else {
+    ares__array_destroy(arr);
```
Comment (tacerus) [inline src/lib/str/ares__buf.c:945]: I'm not sure I'm missing something in my calling code, but trying to use `ares__buf_split_str()` like in the example at https://github.com/c-ares/c-ares/pull/856#issuecomment-2308956347 I get a `free(): double free detected in tcache 2` at this stage.

Comment (bradh352) [reply src/lib/str/ares__buf.c:945]: I need to add some tests still before I merge this.  Infact, this hasn't been tested at all yet, so its not unlikely there's some issues hiding out in this PR.

Comment (tacerus) [reply src/lib/str/ares__buf.c:945]: Ah alright, I'll stay tuned then. :)

Comment (bradh352) [reply src/lib/str/ares__buf.c:951]: ok, fixed now :)

**Pull request merged**